function-calling-sql/function-calling-sql.ipynb

```diff
@@ -0,0 +1,725 @@
+       "<IPython.core.display.Markdown object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "data": {
+      "text/html": [
+       "<div>\n",
+       "<style scoped>\n",
+       "    .dataframe tbody tr th:only-of-type {\n",
+       "        vertical-align: middle;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe tbody tr th {\n",
+       "        vertical-align: top;\n",
+       "    }\n",
+       "\n",
+       "    .dataframe thead th {\n",
+       "        text-align: right;\n",
+       "    }\n",
+       "</style>\n",
+       "<table border=\"1\" class=\"dataframe\">\n",
+       "  <thead>\n",
+       "    <tr style=\"text-align: right;\">\n",
+       "      <th></th>\n",
+       "      <th>product_name</th>\n",
+       "      <th>amount</th>\n",
+       "      <th>name</th>\n",
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "Empty DataFrame\n",
+       "Columns: [product_name, amount, name]\n",
+       "Index: []"
+      ]
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "user_question = 'What are the most recent purchases?'\n",
+    "\n",
+    "\n",
+    "llm_response = text_to_sql(client,system_prompt,user_question,model)\n",
+    "sql_json = json.loads(llm_response)\n",
+    "parsed_sql = sqlparse.format(sql_json['sql'], reindent=True, keyword_case='upper')\n",
+    "formatted_sql = f\"```sql\\n{parsed_sql.strip()}\\n```\"\n",
+    "display(Markdown(formatted_sql)) \n",
+    "\n",
+    "execute_duckdb_query(parsed_sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ea8d3d82-9abf-48b7-8ad4-acb6d38a0310",
+   "metadata": {},
+   "source": [
+    "Obviously, this is not ideal for making any kind of data-driven decisions. It's hard enough to land on a reliable source-of-truth data model, and even harder when your AI analyst cannot give you a consistent result. While text-to-SQL can be great for generating ad-hoc insights, the non-determinism feature of LLMs makes raw text-to-SQL an inpractical solution for a production environment."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cb3918e7-2f0b-4d95-8377-8742c3833a4f",
+   "metadata": {},
+   "source": [
+    "### Function Calling for Verified Queries"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "eb224954-e17a-408c-954e-30757ade35f0",
+   "metadata": {},
+   "source": [
+    "A different approach is to leverage the LLM to call on pre-vetted queries that can answer a set of questions. Since you wouldn't trust a traditional BI tool without rigorously developed and validated SQL, a \"chat with your data\" app should be no different. For this example, we will use the verified queries stored [here](https://github.com/groq/groq-api-cookbook/tree/main/function-calling-sql/verified-queries)."
```

Comment:
```suggestion
    "A different approach is to leverage the LLM to call on pre-vetted queries that can answer a set of questions. Since you wouldn't trust a traditional business intelligence tool without rigorously developed and validated SQL, a \"chat with your data\" app should be no different. For this example, we will use the verified queries stored [here](https://github.com/groq/groq-api-cookbook/tree/main/function-calling-sql/verified-queries)."
```